packages/jest-worker/src/index.js

```diff
@@ -117,95 +108,68 @@ export default class {
       }
 
       // $FlowFixMe: dynamic extension of the class instance is expected.
-      this[name] = this._makeCall.bind(this, name);
+      this[name] = this._callFunctionWithArgs.bind(this, name);
     });
-
-    this._stdout = stdout;
-    this._stderr = stderr;
-    this._ending = false;
-    this._cacheKeys = Object.create(null);
-    this._options = options;
-    this._workers = workers;
-    this._offset = 0;
-  }
-
-  getStdout(): Readable {
-    return this._stdout;
-  }
-
-  getStderr(): Readable {
-    return this._stderr;
-  }
-
-  end() {
-    if (this._ending) {
-      throw new Error('Farm is ended, no more calls can be done to it');
-    }
-
-    const workers = this._workers;
-
-    // We do not cache the request object here. If so, it would only be only
-    // processed by one of the workers, and we want them all to close.
-    for (let i = 0; i < workers.length; i++) {
-      workers[i].send([CHILD_MESSAGE_END, false], emptyMethod, emptyMethod);
-    }
-
-    this._ending = true;
   }
 
   // eslint-disable-next-line no-unclear-flowtypes
-  _makeCall(method: string, ...args: Array<any>): Promise<any> {
+  _callFunctionWithArgs(method: string, ...args: Array<any>): Promise<any> {
     if (this._ending) {
       throw new Error('Farm is ended, no more calls can be done to it');
     }
 
     return new Promise((resolve, reject) => {
       const {computeWorkerKey} = this._options;
-      const workers = this._workers;
-      const length = workers.length;
-      const cacheKeys = this._cacheKeys;
-      const request = [CHILD_MESSAGE_CALL, false, method, args];
+      const request: ChildMessage = [CHILD_MESSAGE_CALL, false, method, args];
 
-      let worker = null;
-      let hash = null;
+      let worker: ?WorkerInterface = null;
+      let hash: ?string = null;
 
       if (computeWorkerKey) {
         hash = computeWorkerKey.apply(this, [method].concat(args));
-        worker = hash == null ? null : cacheKeys[hash];
+        worker = hash == null ? null : this._cacheKeys[hash];
       }
 
-      // Do not use a fat arrow since we need the "this" value, which points to
-      // the worker that executed the call.
-      const onProcessStart = worker => {
+      const onStart: onStart = (worker: WorkerInterface) => {
         if (hash != null) {
-          cacheKeys[hash] = worker;
+          this._cacheKeys[hash] = worker;
         }
       };
 
-      const onProcessEnd = (error, result) => {
+      const onEnd: onEnd = (
+        error: Error,
+        result: mixed,
+        worker: WorkerInterface,
+      ) => {
         if (error) {
           reject(error);
         } else {
           resolve(result);
         }
       };
 
-      // If a worker is pre-selected, use it...
-      if (worker) {
-        worker.send(request, onProcessStart, onProcessEnd);
-        return;
-      }
+      const task = {onEnd, onStart, request};
+      const workerId = worker ? worker.getWorkerId() : undefined;
```

Comment:
Let's unify the interface. Either we work with worker instances or with worker ids, but I see a mix in the API.

Reply:
Done. We use `workerId`s everywhere up to the pool